Polarity based direction heuristics

anwu1219 opened the pull request:
Add method to compute the polarity of a Relu and bias the fixing and case split order based on it.
Tested on 60 acas benchmarks and 40 MNIST benchmarks. 
Solved 11 more SAT and 3 more UNSAT with a 20 minutes timeout compared to the current master.

ahmed-irfan commented on src/configuration/GlobalConfiguration.h:
```diff
@@ -21,6 +21,8 @@ class GlobalConfiguration
 public:
     static void print();
 
+    static const bool USE_POLARITY_BASED_DIRECTION_HEURISTICS;
```
please add some description

anwu1219 replied:
Added some description.

ahmed-irfan commented on src/engine/Engine.cpp:
```diff
@@ -1924,6 +1925,15 @@ void Engine::checkOverallProgress()
+        for ( const auto &constraint : _plConstraints )
+            if ( constraint->supportPolarity() &&
+                 constraint->isActive() && !constraint->phaseFixed() )
+                ( (ReluConstraint *) constraint )->updateDirection();
```
here you are making the assumption that if the plconstraint supports polarity then it must be a reluconstraint. right?

anwu1219 replied:
Added updateDirection as a virtual method in PiecewiseLinearConstraint.h that does nothing by default.

ahmed-irfan commented on src/engine/ReluConstraint.cpp:
```diff
@@ -275,14 +276,30 @@ List<PiecewiseLinearConstraint::Fix> ReluConstraint::getPossibleFixes() const
     {
-        fixes.append( PiecewiseLinearConstraint::Fix( _b, 0 ) );
-        fixes.append( PiecewiseLinearConstraint::Fix( _f, bValue ) );
+        if ( _direction == PHASE_ACTIVE )
```
this and the above code could be merged by rewriting the parent if-else block

ahmed-irfan replied:
after a discussion with @anwu1219 , it seems what I had in my mind is not possible.

ahmed-irfan commented on src/engine/ReluConstraint.cpp:
```diff
@@ -415,6 +432,19 @@ List<PiecewiseLinearCaseSplit> ReluConstraint::getCaseSplits() const
+        splits.append( getActiveSplit() );
+        return splits;
+    }
+    else if ( _direction == PHASE_ACTIVE )
```
remove else

ahmed-irfan commented on src/engine/ReluConstraint.cpp:
```diff
@@ -829,6 +864,21 @@ unsigned ReluConstraint::getAux() const
+{
+    double currentLb = _lowerBounds[_b];
+    double currentUb = _upperBounds[_b];
+    if ( currentLb >= 0 || currentUb <= 0 ) return 1;
```
why currentUb <= 0?

ahmed-irfan replied:
is it `if ( currentLb >= 0 && currentUb <= 0 )` ?

anwu1219 replied:
Whoops this is a bug. Fixed it and added more explanation in the header.

ahmed-irfan commented on src/engine/ReluConstraint.cpp:
```diff
@@ -829,6 +864,21 @@ unsigned ReluConstraint::getAux() const
+    if ( currentLb >= 0 || currentUb <= 0 ) return 1;
+    double width = currentUb - currentLb;
+    double sum = currentUb + currentLb;
+    return sum / width;
```
do we really need to divide the sum by width? isn't sum enough?

anwu1219 replied:
We just need the sign for direction heuristic. However, we need this to compute how symmetric the bound is with respect to zero in the relu-splitting branching heuristic for DnC. Concretely, we will pick ReLUs with the polarity that is closest to 0.

guykatzz commented:
Hi Andrew,
I also reviewed the changes. I think Ahmed raised some good points. In addition:
- When the new polarity flag is off, does the ReLU constraint have the same logic as before?
- We should definitely remove the relu constraint cast in Engine.cpp that Ahmed commented on
- The way polarity is computed is not intuitive; can you perhaps add a short explanation somewhere? Or was this decided empirically?

anwu1219 commented:
@guykatzz 
- yes, when that option is off, the ReLU constraint has the same logic as before.
- Addressed.
- I added a description of how polarity is computed in the ReLuConstraint header file.

ahmed-irfan commented:
@anwu1219 maybe you forgot to push your commits

anwu1219 commented:
Solves 11 more SAT and 3 more UNSAT

guykatzz commented on src/engine/PiecewiseLinearConstraint.h:
```diff
@@ -203,6 +203,23 @@ class PiecewiseLinearConstraint : public ITableau::VariableWatcher
+    }
+
+    /*
+      Return true if and only if this piecewise linear constraint supports
```
Comment seems wrong

anwu1219 replied:
Corrected the comment

guykatzz commented on src/engine/ReluConstraint.cpp:
```diff
@@ -829,6 +864,22 @@ unsigned ReluConstraint::getAux() const
+    double currentLb = _lowerBounds[_b];
+    double currentUb = _upperBounds[_b];
+    if ( currentLb >= 0 ) return 1;
+    if (  currentUb <= 0 ) return -1;
```
remove extra space

guykatzz commented on src/engine/ReluConstraint.h:
```diff
@@ -162,12 +162,42 @@ class ReluConstraint : public PiecewiseLinearConstraint
+      We divide the sum by the width of the interval so that the polarity is
+      always between -1 and 1. The closer it is to 0, the more symmetric the
+      bound is.
+
```
remove empty line

guykatzz commented on src/engine/ReluConstraint.h:
```diff
@@ -162,12 +162,42 @@ class ReluConstraint : public PiecewiseLinearConstraint
+      Update the preferring direction for fixing and handling case split
+    */
+    void updateDirection();
+
```
remove empty line

guykatzz commented on src/engine/ReluConstraint.h:
```diff
@@ -162,12 +162,42 @@ class ReluConstraint : public PiecewiseLinearConstraint
+    */
+    PhaseStatus _direction;
+
+
```
remove empty line

guykatzz commented on src/engine/ReluConstraint.h:
```diff
@@ -162,12 +162,42 @@ class ReluConstraint : public PiecewiseLinearConstraint
+    double computePolarity() const;
+
+    /*
+      Update the preferring direction for fixing and handling case split
```
preferring --> preferred?
